### DISCUSSION TITLE
Filebeat does not release file descriptors on file delete

### DISCUSSION THREAD

**Discussion by trocade** [Security-Onion-Solutions/securityonion#3852] [Unsupported Versions]
So one two of my systems which are under very heavy load suffer from full disk even thou the active used space is only about 500GB at the moment, looking into this show that filebeat is not releasing the deleted file descriptors:

Has this been observerved by anyone else? I would assume there could be a race condition if I look at some old filebeat bug reports.
```
filebeat  16342                  root   45r      REG               8,32 10836536256  147333774 /nsm/zeek/spool/logger/conn.2021-04-08-02-00-00.log (deleted)
filebeat  16342                  root   46r      REG               8,32 12651062010  146672766 /nsm/zeek/spool/logger/conn.2021-04-06-16-00-00.log (deleted)
filebeat  16342                  root   47r      REG               8,32 10307131607  147196189 /nsm/zeek/spool/logger/dns.2021-04-07-13-00-00.log (deleted)
filebeat  16342                  root   48r      REG               8,32 10209237125  146998824 /nsm/zeek/spool/logger/dns.2021-04-07-10-00-00.log (deleted)
filebeat  16342                  root   49r      REG               8,32 10115481124  147587229 /nsm/zeek/spool/logger/dns.2021-04-08-06-00-00.log (deleted)
filebeat  16342                  root   50r      REG               8,32  8826083553  147333922 /nsm/zeek/spool/logger/dns.2021-04-08-04-00-00.log (deleted)
filebeat  16342                  root   52r      REG               8,32  8298407934  146806999 /nsm/zeek/spool/logger/dns.2021-04-09-19-00-00.log (deleted)
filebeat  16342                  root   53r      REG               8,32 13276236323  146806987 /nsm/zeek/spool/logger/conn.2021-04-09-13-00-00.log (deleted)
filebeat  16342                  root   54r      REG               8,32  8813473228  146802255 /nsm/zeek/spool/logger/dns.2021-04-06-18-00-00.log (deleted)
filebeat  16342                  root   55r      REG               8,32 12624168660  146802231 /nsm/zeek/spool/logger/conn.2021-04-06-18-00-00.log (deleted)
filebeat  16342                  root   56r      REG               8,32 10903333256  147333494 /nsm/zeek/spool/logger/conn.2021-04-08-01-00-00.log (deleted)
```
At first I thought it was due to heavy load but the same behaviour can be seen on other sensors that are more or less idle.

Running seconion 2.3.40

**Comment by zambo99** [comment]
Indeed, we are observing the same issue. As soon as filebeat is restarted disk usage goes down again.
We are running securityonion 2.3.2

**Comment by trocade** [comment]
This is a "feature" of filebeat than can be mitigated with this:
https://www.elastic.co/guide/en/beats/filebeat/current/faq-deleted-files-are-not-freed.html

**Reply by zambo99** [reply]
Thanks for your reply, I was already aware of that option.
While I'm convinced that a close_timeout would prevent files from remaining open forever, I still don't understand why this would happen.
In our case the filebeat traffic between the sensor node and the manager node is in the order of the Kbps, and it is not even on all the times (we see intervals of 2-3 minutes in which there is no traffic at all). Therefore, why would filebeat be unable to reach the end-of-file while processing the file?
I'm hesitant to set the close_timeout and forget about it, because this could mean silently dropping logs.

**Comment by TOoSmOotH** [comment]
Maybe start with close_inactive and close_removed first. See if that frees them up.

**Comment by trocade** [comment]
I will try, I did see that the config for filebeat is this for zeek:
```
- type: log
  paths:
    - /nsm/zeek/logs/current/dnp3.log
  fields:
      module: zeek
      dataset: dnp3
      category: network
  processors:
  - drop_fields:
      fields: ["source", "prospector", "input", "offset", "beat"]

  fields_under_root: true
  clean_removed: true
  close_removed: false
```
In the documentation for filebeat it says if close_removed disabled you can not have clean_removed set to true.


But I will add both these options, it is a bit random when it happens in our environment so not sure exacly what triggers the failure to keep going.

**Reply by EbolaWare** [reply]
Any luck on this?

We are running into a similar issue now, but specifically related to logs in `/nsm/zeek/spool/tmp/`. We unfortunately are suffering from high rates of zeek crashing. (It probably should be noted that we are also not completely up to date.) When zeek crashes, it moves logs to that ( `/nsm/zeek/spool/tmp/`) directory, and continues to keep the filebeat harvester open on them indefinitely. 

Currently, I'm weighing options on `ignore_older`, `close_inactive`, `close_renamed`, `close_removed`, `close_timeout`, `clean_inactive` and clean_removed`. _[EDIT: sorry for waking this discussion after a year...]_ I would rather not wind up with gaps in my data, so if you have recommendations based on your experience, I'd appreciate it!

**Reply by pboosten** [reply]
In our case closed_removed and clean_removed to true solved the problems.
